### PR TITLE
[MS] Adding DNS PTR Record

### PR DESCRIPTION
**Pull request by echuvyrov**
Enabling the creation of DNS PTR Record

### CODE REVIEW

File: azurerm/resource_arm_dns_ptr_record.go [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+			"zone_name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
+
+			"records": &schema.Schema{
+				Type:     schema.TypeSet,
+				Required: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
```
Comment (whiskeyjay) [inline azurerm/resource_arm_dns_ptr_record.go]: Can this contain zero records?

File: azurerm/resource_arm_dns_ptr_record.go [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+				Type:     schema.TypeInt,
+				Required: true,
+			},
+
+			"tags": tagsSchema(),
+		},
+	}
+}
+
+func resourceArmDnsPtrRecordCreate(d *schema.ResourceData, meta interface{}) error {
```
Comment (whiskeyjay) [inline azurerm/resource_arm_dns_ptr_record.go]: resourceArmDnsPtrRecordCreateOrUpdate

File: azurerm/resource_arm_dns_ptr_record.go [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+		TTL:        &ttl,
+		PtrRecords: &records,
+	}
+
+	parameters := dns.RecordSet{
+		Name:                &name,
+		RecordSetProperties: &props,
+	}
+
+	_, err := dnsClient.CreateOrUpdate(resGroup, zoneName, name, dns.PTR, parameters, "", "")
```
Comment (whiskeyjay) [inline azurerm/resource_arm_dns_ptr_record.go]: It is not obvious what the last two parameters are and why are they empty strings.

File: azurerm/resource_arm_dns_ptr_record.go [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+	if err != nil {
+		return err
+	}
+
+	rec, err := dnsClient.Get(resGroup, zoneName, name, dns.PTR)
```
Comment (whiskeyjay) [inline azurerm/resource_arm_dns_ptr_record.go]: I believe the dnsClient.CreateOrUpdate method's "response" return value already contains the resource ID, there is no need to call Get just to retrieve the ID value.

```
rec, err := dnsClient.CreateOrUpdate(...)

if err != nil {
    return err
}
d.SetId(*rec.ID)
```

File: azurerm/provider.go [hunk 5]
```diff
@@ -124,6 +124,8 @@ func Provider() terraform.ResourceProvider {
 
 			"azurerm_application_insights": resourceArmApplicationInsights(),
 
+			//PTR record uses Azure Go SDK
```
Comment (tombuildsstuff) [inline azurerm/provider.go]: _minor_: can we move this sorted alphabetically under the Azure SDK for Go comment?

File: azurerm/resource_arm_dns_ptr_record.go [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+
+	name := d.Get("name").(string)
+	resGroup := d.Get("resource_group_name").(string)
+	zoneName := d.Get("zone_name").(string)
+	ttl := int64(d.Get("ttl").(int))
+	tags := d.Get("tags").(map[string]interface{})
+
+	metadata := expandTags(tags)
+
+	recordStrings := d.Get("records").(*schema.Set).List()
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_dns_ptr_record.go]: we'd generally tend to pull this out into a separate function - could we update this to match?

File: azurerm/resource_arm_dns_ptr_record.go [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+		return err
+	}
+
+	resGroup := id.ResourceGroup
+	name := id.Path["PTR"]
+	zoneName := id.Path["dnszones"]
+
+	resp, err := dnsClient.Get(resGroup, zoneName, name, dns.PTR)
+	if err != nil {
+		return fmt.Errorf("Error reading DNS PTR record %s: %s", name, err)
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_dns_ptr_record.go]: this last one should be `%+v` as it's an error

File: azurerm/resource_arm_dns_ptr_record.go [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+		d.SetId("")
+		return nil
+	}
+
+	d.Set("name", name)
+	d.Set("resource_group_name", resGroup)
+	d.Set("zonename", zoneName)
+	d.Set("ttl", resp.TTL)
+
+	if resp.PtrRecords != nil {
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_dns_ptr_record.go]: we'd generally pull this out into a Flatten method - could we update this to match?

File: azurerm/resource_arm_dns_ptr_record_test.go [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+
+	"github.com/Azure/azure-sdk-for-go/arm/dns"
+	"github.com/hashicorp/terraform/helper/acctest"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/terraform"
+)
+
+func TestAccAzureRMDnsPtrRecord_basic(t *testing.T) {
+	ri := acctest.RandInt()
+	config := fmt.Sprintf(testAccAzureRMDnsPtrRecord_basic, ri, ri, ri)
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_dns_ptr_record_test.go]: could we switch this to match the newer resources by making this a function which returns a formatted string?

File: azurerm/resource_arm_dns_ptr_record_test.go [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+				),
+			},
+		},
+	})
+}
+
+func TestAccAzureRMDnsPtrRecord_updateRecords(t *testing.T) {
+	ri := acctest.RandInt()
+	preConfig := fmt.Sprintf(testAccAzureRMDnsPtrRecord_basic, ri, ri, ri)
+	postConfig := fmt.Sprintf(testAccAzureRMDnsPtrRecord_updateRecords, ri, ri, ri)
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_dns_ptr_record_test.go]: (same here)

File: azurerm/resource_arm_dns_ptr_record_test.go [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+				),
+			},
+		},
+	})
+}
+
+func TestAccAzureRMDnsPtrRecord_withTags(t *testing.T) {
+	ri := acctest.RandInt()
+	preConfig := fmt.Sprintf(testAccAzureRMDnsPtrRecord_withTags, ri, ri, ri)
+	postConfig := fmt.Sprintf(testAccAzureRMDnsPtrRecord_withTagsUpdate, ri, ri, ri)
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_dns_ptr_record_test.go]: (same here)

File: azurerm/resource_arm_dns_ptr_record_test.go [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testCheckAzureRMDnsPtrRecordDestroy,
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config: preConfig,
+				Check: resource.ComposeTestCheckFunc(
+					testCheckAzureRMDnsPtrRecordExists("azurerm_dns_ptr_record.test"),
+					resource.TestCheckResourceAttr(
+						"azurerm_dns_ptr_record.test", "tags.%", "2"),
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_dns_ptr_record_test.go]: do we want to check the ordering of the values here?

File: azurerm/resource_arm_dns_ptr_record_test.go [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+		zoneName := rs.Primary.Attributes["zone_name"]
+		resourceGroup, hasResourceGroup := rs.Primary.Attributes["resource_group_name"]
+		if !hasResourceGroup {
+			return fmt.Errorf("Bad: no resource group found in state for DNS PTR record: %s", ptrName)
+		}
+
+		conn := testAccProvider.Meta().(*ArmClient).dnsClient
+		resp, err := conn.Get(resourceGroup, zoneName, ptrName, dns.PTR)
+		if err != nil {
+			return fmt.Errorf("Bad: Get PTR RecordSet: %s", err)
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_dns_ptr_record_test.go]: this needs to be `%+v`

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
Hey @echuvyrov 

Thanks for this PR - I've taken a look and left some comments in-line - but this is looking good!

Thanks!

File: website/docs/r/dns_ptr_record.html.markdown [hunk 14]
```diff
@@ -0,0 +1,63 @@
+---
+layout: "azurerm"
+page_title: "Azure Resource Manager: azurerm_dns_ptr_record"
+sidebar_current: "docs-azurerm-resource-dns-ptr-record"
+description: |-
+  Create a DNS PTR Record.
+---
+
+# azurerm\_dns\_a\_record
```
Comment (tombuildsstuff) [inline website/docs/r/dns_ptr_record.html.markdown]: *`ptr`

File: azurerm/resource_arm_dns_ptr_record.go (hunk 1 above)
Comment (echuvyrov) [inline azurerm/resource_arm_dns_ptr_record.go]: It cannot, it would be invalid

File: azurerm/resource_arm_dns_ptr_record.go (hunk 2 above)
Comment (echuvyrov) [inline azurerm/resource_arm_dns_ptr_record.go]: changed

File: azurerm/resource_arm_dns_ptr_record.go (hunk 3 above)
Comment (echuvyrov) [inline azurerm/resource_arm_dns_ptr_record.go]: good catch, changed second-to-last parameter to properly be the eTag of the record and added comments explaining why the last parameter is empty

File: azurerm/resource_arm_dns_ptr_record.go (hunk 4 above)
Comment (echuvyrov) [inline azurerm/resource_arm_dns_ptr_record.go]: good catch, changed

File: azurerm/provider.go (hunk 5 above)
Comment (echuvyrov) [inline azurerm/provider.go]: done

File: azurerm/resource_arm_dns_ptr_record.go (hunk 6 above)
Comment (echuvyrov) [inline azurerm/resource_arm_dns_ptr_record.go]: changed

File: azurerm/resource_arm_dns_ptr_record.go (hunk 7 above)
Comment (echuvyrov) [inline azurerm/resource_arm_dns_ptr_record.go]: changed, thanks for catching that!

File: azurerm/resource_arm_dns_ptr_record.go (hunk 8 above)
Comment (echuvyrov) [inline azurerm/resource_arm_dns_ptr_record.go]: changed

File: azurerm/resource_arm_dns_ptr_record_test.go (hunk 9 above)
Comment (echuvyrov) [inline azurerm/resource_arm_dns_ptr_record_test.go]: changed

File: azurerm/resource_arm_dns_ptr_record_test.go (hunk 10 above)
Comment (echuvyrov) [inline azurerm/resource_arm_dns_ptr_record_test.go]: changed

File: azurerm/resource_arm_dns_ptr_record_test.go (hunk 11 above)
Comment (echuvyrov) [inline azurerm/resource_arm_dns_ptr_record_test.go]: changed

File: azurerm/resource_arm_dns_ptr_record_test.go (hunk 12 above)
Comment (echuvyrov) [inline azurerm/resource_arm_dns_ptr_record_test.go]: changed

File: azurerm/resource_arm_dns_ptr_record_test.go (hunk 13 above)
Comment (echuvyrov) [inline azurerm/resource_arm_dns_ptr_record_test.go]: changed

File: website/docs/r/dns_ptr_record.html.markdown (hunk 14 above)
Comment (echuvyrov) [inline website/docs/r/dns_ptr_record.html.markdown]: changed

**Comment by echuvyrov** [comment]
@tombuildsstuff @whiskeyjay made the changes you guys suggested, thanks! The only one I wasn't sure about was about order checking in the tests - I added tag checking, lmk if you meant something else.

**Review by tombuildsstuff — APPROVED** [review]
Hey @echuvyrov

Thanks for pushing those updates - I've pushed a commit to resolve a couple of minor issues I noticed below (I hope you don't mind) - and this now LGTM :)

Tests pass:
```
$ TF_ACC=1 envchain azurerm go test ./azurerm -v -timeout 120m -run TestAccAzureRMDns
=== RUN   TestAccAzureRMDnsARecord_importBasic
--- PASS: TestAccAzureRMDnsARecord_importBasic (106.92s)
=== RUN   TestAccAzureRMDnsAAAARecord_importBasic
--- PASS: TestAccAzureRMDnsAAAARecord_importBasic (92.47s)
=== RUN   TestAccAzureRMDnsCNameRecord_importBasic
--- PASS: TestAccAzureRMDnsCNameRecord_importBasic (92.02s)
=== RUN   TestAccAzureRMDnsMxRecord_importBasic
--- PASS: TestAccAzureRMDnsMxRecord_importBasic (91.66s)
=== RUN   TestAccAzureRMDnsNsRecord_importBasic
--- PASS: TestAccAzureRMDnsNsRecord_importBasic (92.20s)
=== RUN   TestAccAzureRMDnsPtrRecord_importBasic
--- PASS: TestAccAzureRMDnsPtrRecord_importBasic (74.45s)
=== RUN   TestAccAzureRMDnsSrvRecord_importBasic
--- PASS: TestAccAzureRMDnsSrvRecord_importBasic (94.21s)
=== RUN   TestAccAzureRMDnsTxtRecord_importBasic
--- PASS: TestAccAzureRMDnsTxtRecord_importBasic (93.43s)
=== RUN   TestAccAzureRMDnsZone_importBasic
--- PASS: TestAccAzureRMDnsZone_importBasic (89.80s)
=== RUN   TestAccAzureRMDnsARecord_basic
--- PASS: TestAccAzureRMDnsARecord_basic (88.74s)
=== RUN   TestAccAzureRMDnsARecord_updateRecords
--- PASS: TestAccAzureRMDnsARecord_updateRecords (91.76s)
=== RUN   TestAccAzureRMDnsARecord_withTags
--- PASS: TestAccAzureRMDnsARecord_withTags (103.81s)
=== RUN   TestAccAzureRMDnsAAAARecord_basic
--- PASS: TestAccAzureRMDnsAAAARecord_basic (78.32s)
=== RUN   TestAccAzureRMDnsAAAARecord_updateRecords
--- PASS: TestAccAzureRMDnsAAAARecord_updateRecords (106.47s)
=== RUN   TestAccAzureRMDnsAAAARecord_withTags
--- PASS: TestAccAzureRMDnsAAAARecord_withTags (110.22s)
=== RUN   TestAccAzureRMDnsCNameRecord_basic
--- PASS: TestAccAzureRMDnsCNameRecord_basic (89.36s)
=== RUN   TestAccAzureRMDnsCNameRecord_subdomain
--- PASS: TestAccAzureRMDnsCNameRecord_subdomain (95.82s)
=== RUN   TestAccAzureRMDnsCNameRecord_updateRecords
--- PASS: TestAccAzureRMDnsCNameRecord_updateRecords (94.08s)
=== RUN   TestAccAzureRMDnsCNameRecord_withTags
--- PASS: TestAccAzureRMDnsCNameRecord_withTags (108.93s)
=== RUN   TestAccAzureRMDnsMxRecord_basic
--- PASS: TestAccAzureRMDnsMxRecord_basic (74.31s)
=== RUN   TestAccAzureRMDnsMxRecord_updateRecords
--- PASS: TestAccAzureRMDnsMxRecord_updateRecords (106.70s)
=== RUN   TestAccAzureRMDnsMxRecord_withTags
--- PASS: TestAccAzureRMDnsMxRecord_withTags (136.45s)
=== RUN   TestAccAzureRMDnsNsRecord_basic
--- PASS: TestAccAzureRMDnsNsRecord_basic (91.50s)
=== RUN   TestAccAzureRMDnsNsRecord_updateRecords
--- PASS: TestAccAzureRMDnsNsRecord_updateRecords (145.28s)
=== RUN   TestAccAzureRMDnsNsRecord_withTags
--- PASS: TestAccAzureRMDnsNsRecord_withTags (106.33s)
=== RUN   TestAccAzureRMDnsPtrRecord_basic
--- PASS: TestAccAzureRMDnsPtrRecord_basic (82.45s)
=== RUN   TestAccAzureRMDnsPtrRecord_updateRecords
--- PASS: TestAccAzureRMDnsPtrRecord_updateRecords (125.35s)
=== RUN   TestAccAzureRMDnsPtrRecord_withTags
--- PASS: TestAccAzureRMDnsPtrRecord_withTags (89.33s)
=== RUN   TestAccAzureRMDnsSrvRecord_basic
--- PASS: TestAccAzureRMDnsSrvRecord_basic (70.97s)
=== RUN   TestAccAzureRMDnsSrvRecord_updateRecords
--- PASS: TestAccAzureRMDnsSrvRecord_updateRecords (103.76s)
=== RUN   TestAccAzureRMDnsSrvRecord_withTags
--- PASS: TestAccAzureRMDnsSrvRecord_withTags (89.54s)
=== RUN   TestAccAzureRMDnsTxtRecord_basic
--- PASS: TestAccAzureRMDnsTxtRecord_basic (73.52s)
=== RUN   TestAccAzureRMDnsTxtRecord_updateRecords
--- PASS: TestAccAzureRMDnsTxtRecord_updateRecords (101.40s)
=== RUN   TestAccAzureRMDnsTxtRecord_withTags
--- PASS: TestAccAzureRMDnsTxtRecord_withTags (104.44s)
=== RUN   TestAccAzureRMDnsZone_basic
--- PASS: TestAccAzureRMDnsZone_basic (85.98s)
=== RUN   TestAccAzureRMDnsZone_withTags
--- PASS: TestAccAzureRMDnsZone_withTags (99.68s)
PASS
ok  	github.com/terraform-providers/terraform-provider-azurerm/azurerm	3481.650s
```

Thanks!

**Pull request merged**